File: fasta/Version3/Claude/fasta3.c
/*
 * The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * Converted from Python to C by Claude
 * Original Python by Ian Osgood, Heinrich Acker, Justin Peel, and Christopher Sean Forgeron
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define LINE_WIDTH 60
#define IM 139968.0

// ALU sequence
const char *alu =
   "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG"
   "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA"
   "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT"
   "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA"
   "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG"
   "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC"
   "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

// Frequency tables
typedef struct {
    char c;
    double p;
} Frequency;

Frequency iub[] = {
    {'a', 0.27}, {'c', 0.12}, {'g', 0.12}, {'t', 0.27},
    {'B', 0.02}, {'D', 0.02}, {'H', 0.02}, {'K', 0.02},
    {'M', 0.02}, {'N', 0.02}, {'R', 0.02}, {'S', 0.02},
    {'V', 0.02}, {'W', 0.02}, {'Y', 0.02}
};

Frequency homosapiens[] = {
    {'a', 0.3029549426680},
    {'c', 0.1979883004921},
    {'g', 0.1975473066391},
    {'t', 0.3015094502008}
};

// Convert frequency table to cumulative probabilities
typedef struct {
    double prob;
    char c;
} CumulativeFreq;

void make_cumulative(Frequency *table, int table_size, CumulativeFreq *result) {
    double prob = 0.0;
    for (int i = 0; i < table_size; i++) {
        prob += table[i].p;
        result[i].prob = prob;
        result[i].c = table[i].c;
    }
}

// Binary search to find character based on probability
char lookup_char(CumulativeFreq *table, int table_size, double random) {
    int low = 0;
    int high = table_size - 1;
    
    while (low <= high) {
        int mid = (low + high) / 2;
        if (random < table[mid].prob)
            high = mid - 1;
        else
            low = mid + 1;
    }
    
    return table[low].c;
}

// Generate a repeated sequence
void repeat_fasta(const char *src, int src_len, int n) {
    char buffer[LINE_WIDTH + 1];
    buffer[LINE_WIDTH] = '\n';
    
    int pos = 0;
    int remaining = n;
    
    while (remaining > 0) {
        int line_len = remaining < LINE_WIDTH ? remaining : LINE_WIDTH;
        for (int i = 0; i < line_len; i++) {
            buffer[i] = src[(pos + i) % src_len];
        }
        fwrite(buffer, 1, line_len + (line_len == LINE_WIDTH ? 1 : 0), stdout);
        if (line_len < LINE_WIDTH) {
            putchar('\n');
        }
        
        pos = (pos + line_len) % src_len;
        remaining -= line_len;
    }
}

// Generate a random sequence
double random_fasta(CumulativeFreq *table, int table_size, int n, double seed) {
    char buffer[LINE_WIDTH + 1];
    buffer[LINE_WIDTH] = '\n';
    
    int remaining = n;
    
    while (remaining > 0) {
        int line_len = remaining < LINE_WIDTH ? remaining : LINE_WIDTH;
        
        for (int i = 0; i < line_len; i++) {
            seed = (seed * 3877.0 + 29573.0) % IM;
            buffer[i] = lookup_char(table, table_size, seed / IM);
        }
        
        fwrite(buffer, 1, line_len + (line_len == LINE_WIDTH ? 1 : 0), stdout);
        if (line_len < LINE_WIDTH) {
            putchar('\n');
        }
        
        remaining -= line_len;
    }
    
    return seed;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <number>\n", argv[0]);
        return 1;
    }
    
    int n = atoi(argv[1]);
    
    // Calculate the size of the frequency tables
    int iub_size = sizeof(iub) / sizeof(iub[0]);
    int homosapiens_size = sizeof(homosapiens) / sizeof(homosapiens[0]);
    
    // Convert the frequency tables to cumulative probabilities
    CumulativeFreq iub_cumul[iub_size];
    CumulativeFreq homosapiens_cumul[homosapiens_size];
    
    make_cumulative(iub, iub_size, iub_cumul);
    make_cumulative(homosapiens, homosapiens_size, homosapiens_cumul);
    
    // Generate the sequences
    printf(">ONE Homo sapiens alu\n");
    repeat_fasta(alu, strlen(alu), n * 2);
    
    printf(">TWO IUB ambiguity codes\n");
    double seed = 42.0;
    seed = random_fasta(iub_cumul, iub_size, n * 3, seed);
    
    printf(">THREE Homo sapiens frequency\n");
    random_fasta(homosapiens_cumul, homosapiens_size, n * 5, seed);
    
    return 0;
}

File: fasta/Version3/Claude/fasta3.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// Converted from Python to Rust by Claude

use std::env;
use std::io::{self, Write, BufWriter};

const ALU: &str = concat!(
    "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG",
    "GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA",
    "CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT",
    "ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA",
    "GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG",
    "AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC",
    "AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"
);

const WIDTH: usize = 60;
const IM: f64 = 139968.0;

// Frequency tables
const IUB: [(&str, f64); 15] = [
    ("a", 0.27), ("c", 0.12), ("g", 0.12), ("t", 0.27),
    ("B", 0.02), ("D", 0.02), ("H", 0.02), ("K", 0.02),
    ("M", 0.02), ("N", 0.02), ("R", 0.02), ("S", 0.02),
    ("V", 0.02), ("W", 0.02), ("Y", 0.02)
];

const HOMOSAPIENS: [(&str, f64); 4] = [
    ("a", 0.3029549426680),
    ("c", 0.1979883004921),
    ("g", 0.1975473066391),
    ("t", 0.3015094502008)
];

// Convert frequency table to cumulative probabilities
fn make_cumulative(table: &[(&str, f64)]) -> (Vec<f64>, Vec<u8>) {
    let mut probs = Vec::with_capacity(table.len());
    let mut chars = Vec::with_capacity(table.len());
    let mut prob = 0.0;
    
    for &(ch, p) in table {
        prob += p;
        probs.push(prob);
        chars.push(ch.as_bytes()[0]);
    }
    
    (probs, chars)
}

// Binary search to find character based on probability
fn bisect(probs: &[f64], r: f64) -> usize {
    let mut low = 0;
    let mut high = probs.len() - 1;
    
    while low < high {
        let mid = (low + high) / 2;
        if r < probs[mid] {
            high = mid;
        } else {
            low = mid + 1;
        }
    }
    
    low
}

// Generate a repeated sequence
fn repeat_fasta<W: Write>(writer: &mut W, src: &str, n: usize) -> io::Result<()> {
    let src_bytes = src.as_bytes();
    let src_len = src.len();
    let mut line = Vec::with_capacity(WIDTH + 1);
    line.resize(WIDTH + 1, 0);
    line[WIDTH] = b'\n';
    
    let mut pos = 0;
    let mut remaining = n;
    
    while remaining > 0 {
        let line_len = remaining.min(WIDTH);
        
        // Fill the line buffer
        for i in 0..line_len {
            line[i] = src_bytes[(pos + i) % src_len];
        }
        
        // Write the line
        if line_len == WIDTH {
            writer.write_all(&line[..WIDTH + 1])?;
        } else {
            line[line_len] = b'\n';
            writer.write_all(&line[..line_len + 1])?;
        }
        
        pos = (pos + line_len) % src_len;
        remaining -= line_len;
    }
    
    Ok(())
}

// Generate a random sequence
fn random_fasta<W: Write>(
    writer: &mut W,
    probs: &[f64],
    chars: &[u8],
    n: usize,
    mut seed: f64
) -> io::Result<f64> {
    let mut line = Vec::with_capacity(WIDTH + 1);
    line.resize(WIDTH + 1, 0);
    line[WIDTH] = b'\n';
    
    let mut remaining = n;
    
    while remaining > 0 {
        let line_len = remaining.min(WIDTH);
        
        for i in 0..line_len {
            // Generate random number
            seed = (seed * 3877.0 + 29573.0) % IM;
            let r = seed / IM;
            
            // Find character based on probability
            let index = bisect(probs, r);
            line[i] = chars[index];
        }
        
        // Write the line
        if line_len == WIDTH {
            writer.write_all(&line[..WIDTH + 1])?;
        } else {
            line[line_len] = b'\n';
            writer.write_all(&line[..line_len + 1])?;
        }
        
        remaining -= line_len;
    }
    
    Ok(seed)
}

fn main() -> io::Result<()> {
    // Parse command-line argument
    let n = env::args()
        .nth(1)
        .unwrap_or_else(|| "1000".to_string())
        .parse::<usize>()
        .expect("Expected a number");
    
    // Use a buffered writer for better performance
    let stdout = io::stdout();
    let mut writer = BufWriter::with_capacity(64 * 1024, stdout.lock());
    
    // Create cumulative frequency tables
    let (iub_probs, iub_chars) = make_cumulative(&IUB);
    let (homosapiens_probs, homosapiens_chars) = make_cumulative(&HOMOSAPIENS);
    
    // Generate sequences
    writeln!(writer, ">ONE Homo sapiens alu")?;
    repeat_fasta(&mut writer, ALU, n * 2)?;
    
    writeln!(writer, ">TWO IUB ambiguity codes")?;
    let seed = random_fasta(&mut writer, &iub_probs, &iub_chars, n * 3, 42.0)?;
    
    writeln!(writer, ">THREE Homo sapiens frequency")?;
    random_fasta(&mut writer, &homosapiens_probs, &homosapiens_chars, n * 5, seed)?;
    
    Ok(())
}